dom.go

```diff
@@ -276,18 +325,105 @@ func (h *HTML) reconcile(prev *HTML) {
+		case nextChildRender != nil && prevChildRender == nil && insertBefore != nil:
+			h.node.Call("insertBefore", nextChildRender.node, insertBefore.node)
+		default:
+			replaceNode(nextChildRender.node, prevChildRender.node)
+		}
+	}
+	h.removeChildren(prev)
+}
+
+// removeChildren removes child elements from the previous render pass that no
+// longer exist on the current HTML children.
+func (h *HTML) removeChildren(prev *HTML) {
+	if prev == nil {
+		return
 	}
 	for i := len(h.children); i < len(prev.children); i++ {
 		prevChild := prev.children[i]
+		if isChildList(prevChild) {
+			prevChild.(ChildList).removeExcept(h.node, -1)
+			continue
+		}
 		prevChildRender := extractHTML(prevChild)
+		if prevChildRender == nil {
+			continue
+		}
 		removeNode(prevChildRender.node)
 		if u, ok := prevChild.(Unmounter); ok {
 			u.Unmount()
 		}
 	}
 }
 
+// ChildList represents a list of Markup, Component, or HTML which is
+// individually applied to an HTML element or text node.
+type ChildList []ComponentOrHTML
+
+// reconcile the ChildList against the DOM node in isolation.  If insertBefore
+// is non-nil, the ChildList elements will be inserted into the DOM before that
+// node.
+func (l ChildList) reconcile(node jsObject, insertBefore *HTML, prev ComponentOrHTML) {
+	nextHTML := &HTML{node: node, children: l}
+	switch c := prev.(type) {
+	case ChildList:
+		nextHTML.reconcileChildren(&HTML{node: node, children: c}, insertBefore)
+	default:
+		if prev == nil {
+			nextHTML.reconcileChildren(&HTML{node: node}, insertBefore)
+		} else {
+			nextHTML.reconcileChildren(&HTML{node: node, children: []ComponentOrHTML{prev}}, insertBefore)
+		}
+	}
+}
+
+// firstHTML returns the index and content of the first element from which a
+// *HTML can be extracted.  Returns an index of -1 if not found.
+func (l ChildList) firstHTML() (idx int, h *HTML) {
+	for idx = 0; idx < len(l); idx++ {
+		if isChildList(l[idx]) {
+			_, h = l[idx].(ChildList).firstHTML()
+		} else {
+			h = extractHTML(l[idx])
+		}
+		if h != nil {
+			return idx, h
+		}
+	}
+	return -1, nil
+}
+
+// removeExcept removes the ChildList's elements from the DOM node, except for
+// the element at index exceptIndex.  If exceptIndex is less than zero, all
+// elements are removed.
+func (l ChildList) removeExcept(node jsObject, exceptIndex int) {
+	var children ChildList
+	switch {
+	case exceptIndex < 0:
+		children = l
+	case isChildList(l[exceptIndex]):
+		// TODO(pdf): I'm not sure I like the assumption here that we want to
+		// retain the first element of sub-lists, but not sure if there's a
+		// better method for handling this.
+		idx, _ := l[exceptIndex].(ChildList).firstHTML()
+		l[exceptIndex].(ChildList).removeExcept(node, idx)
+		children = append(l[:exceptIndex], l[exceptIndex+1:]...)
+	default:
+		children = append(l[:exceptIndex], l[exceptIndex+1:]...)
+	}
+	nextHTML := &HTML{node: node}
+	nextHTML.removeChildren(&HTML{node: node, children: children})
+}
+
```

Comment:
All of the code in this file above this line is quite complex, and I don't understand most of it currently (I've only had ~15 mins to look at it, my lunch break is up now).

I'll have more time to look at this later today or tomorrow, but I think we should be able to simplify a lot of this or make it more clear what is going on. Hopefully my other comments will at least let us solidify the public API.

Reply:
Thanks @slimsag, open to suggestions - as I said in the OP, I do not expect this to be final code, but a first functional pass.  The list handling is necessarily somewhat complex though.

Reply:
Understandable :) I'm going to take a more deep look at this code tonight or tomorrow and offer more concrete suggestions than just "the code is complex" :P I think this PR is a really good starting point!